File: hush-core/hush/core/tracing/local.py
"""LocalTracer — writes trace data to local JSON files.

Zero dependencies, zero setup. One JSON file per request.
Default path: ~/.hush/traces/ (configurable via path= or HUSH_TRACES_DIR env var).
"""

import json
import logging
import os
from pathlib import Path
from typing import Any, Dict, List, Optional

from hush.core.tracing.base import Tracer

LOGGER = logging.getLogger("hush.tracing")


class LocalTracer(Tracer):
    """Tracer that writes trace data to local JSON files.

    Each engine.run() produces one file: {path}/{request_id}.json
    containing the full TracePayload as pretty-printed JSON.

    Example:
        from hush.core.tracing import LocalTracer

        engine = Hush(graph)
        result = await engine.run(inputs, tracer=LocalTracer())
        # Trace written to ~/.hush/traces/{request_id}.json

        # Custom path
        result = await engine.run(inputs, tracer=LocalTracer(path="./my-traces"))
    """

    def __init__(
        self,
        path: Optional[str] = None,
        tags: Optional[List[str]] = None,
    ):
        super().__init__(tags=tags, stream_trace_limit=None)
        self._path = Path(
            path or os.environ.get("HUSH_TRACES_DIR", "~/.hush/traces")
        ).expanduser()

    def flush(self, trace_data: Dict[str, Any]) -> None:
        """Write trace data to a JSON file.

        Args:
            trace_data: Dict matching TracePayload format.
        """
        self._path.mkdir(parents=True, exist_ok=True)
        request_id = trace_data.get("request_id", "unknown")
        filepath = self._path / f"{request_id}.json"
        filepath.write_text(json.dumps(trace_data, indent=2, default=str))
        LOGGER.info("Trace written to %s", filepath)

    def __repr__(self) -> str:
        return f"<LocalTracer path={self._path}>"

File: rust/hush-icore/src/tracing/local.rs
//! LocalTracer — writes trace data to local JSON files.
//!
//! Mirrors Python's `hush.core.tracing.local.LocalTracer`.
//! Zero dependencies, zero setup. One JSON file per request.
//! Default path: ~/.hush/traces/ (configurable via path or HUSH_TRACES_DIR env var).

use std::fmt;
use std::fs;
use std::path::PathBuf;

use serde_json::Value;

use crate::tracing::tracer::Tracer;

/// Get home directory using standard env vars (no external crate).
fn home_dir() -> PathBuf {
    std::env::var("HOME")
        .or_else(|_| std::env::var("USERPROFILE"))
        .map(PathBuf::from)
        .unwrap_or_else(|_| PathBuf::from("."))
}

/// Tracer that writes trace data to local JSON files.
///
/// Each engine.run() produces one file: `{path}/{request_id}.json`
/// containing the trace data as pretty-printed JSON.
pub struct LocalTracer {
    path: PathBuf,
    tags: Vec<String>,
}

impl LocalTracer {
    /// Create a new LocalTracer.
    ///
    /// `path`: Directory for trace files. Defaults to `~/.hush/traces/`
    /// or the `HUSH_TRACES_DIR` environment variable.
    pub fn new(path: Option<&str>, tags: Vec<String>) -> Self {
        let dir = path
            .map(PathBuf::from)
            .or_else(|| std::env::var("HUSH_TRACES_DIR").ok().map(PathBuf::from))
            .unwrap_or_else(|| {
                home_dir().join(".hush").join("traces")
            });
        LocalTracer { path: dir, tags }
    }
}

impl fmt::Debug for LocalTracer {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "<LocalTracer path={}>", self.path.display())
    }
}

impl Tracer for LocalTracer {
    fn flush(&self, trace_data: Value) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        fs::create_dir_all(&self.path)?;
        let request_id = trace_data
            .get("request_id")
            .and_then(|v| v.as_str())
            .unwrap_or("unknown");
        let filepath = self.path.join(format!("{request_id}.json"));
        let json_str = serde_json::to_string_pretty(&trace_data)?;
        fs::write(&filepath, json_str)?;
        eprintln!("[hush-tracing] Trace written to {}", filepath.display());
        Ok(())
    }

    fn tags(&self) -> Vec<String> {
        self.tags.clone()
    }

    fn stream_trace_limit(&self) -> Option<usize> {
        None // Keep all traces for local debugging
    }
}
